internal/gps/source_manager.go

```diff
@@ -455,6 +462,72 @@ func (sm *SourceMgr) DeduceProjectRoot(ip string) (ProjectRoot, error) {
 	return ProjectRoot(pd.root), err
 }
 
+// DeduceConstraint tries to puzzle out what kind of version is given in a string -
+// semver, a revision, or as a fallback, a plain tag
+func (sm *SourceMgr) DeduceConstraint(s string, pi ProjectIdentifier) (Constraint, error) {
```

Comment:
Let's rename this to `InferConstraint`. This is especially important given the below plan about changing how we do revision handling; before we were doing some deduction and inference, but after we make those changes to revision handling, eventually it'll really just be inference.